### ISSUE TITLE
Question about "Required, can be None" parameter

### ISSUE THREAD

**Issue by Kludex** [fastapi/fastapi#12419]
### Discussed in https://github.com/fastapi/fastapi/discussions/12418

<div type='discussions-op-text'>

<sup>Originally posted by **bowenerchen** October 10, 2024</sup>
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/items/")
async def read_items(q: Annotated[str | None, Query(min_length=3)] = ...):
    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q:
        results.update({"q": q})
    return results
```

### Description

https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#required-can-be-none
![QQ_1728534861112](https://github.com/user-attachments/assets/95e967aa-ea9f-46c1-80c8-f13dc7c52e6c)
how to pass a None value to the required query parameter q?

### Operating System

macOS

### Operating System Details

Sonoma 14.7 M2 Max

### FastAPI Version

0.115.0

### Pydantic Version

2.5.3

### Python Version

Python 3.12.4

### Additional Context

_No response_</div>

**Comment by Kludex** [comment]
The documentation is wrong.

This is the file that is testing that part of the documentation: https://github.com/fastapi/fastapi/blob/529155e72e196f82e0289f173a185e8ddb2888ce/tests/test_tutorial/test_query_params/test_tutorial006_py310.py

As we can see, it doesn't pass the `None`...

**Comment by Kludex** [comment]
This may be the same as https://github.com/fastapi/fastapi/issues/12245

**Comment by zubairevangelist** [comment]
Hi @Kludex, should i open a PR to `improve `documentation`` like below one, this will be my first open source contribution?

```
@app.get("/items/")
async def read_items(q: Annotated[str | None, Query()]):  # required
    if q == "null":
        q = None

    results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
    if q is not None:
        results.update({"q": q})
    return results
```

**Comment by WilliamDEdwards** [comment]
> Hi [@Kludex](https://github.com/Kludex), should i open a PR to `improve documentation` like below one, this will be my first open source contribution?
> 
> ```
> @app.get("/items/")
> async def read_items(q: Annotated[str | None, Query()]):  # required
>     if q == "null":
>         q = None
> 
>     results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
>     if q is not None:
>         results.update({"q": q})
>     return results
> ```

"Don't ask for permission, ask for forgiveness"

**Comment by khuloud0** [comment]
Hi everyone, I've reproduced this on my local machine. When using Annotated[Union[str, None], Query(min_length=3)] = ..., the Swagger UI marks the field as Required and won't allow execution if it's empty. This confirms the confusion mentioned in the 'Required, can be None' section of the docs. I'm interested in helping fix this in the documentation. Should we update the table to clarify that ... always renders as required in the UI regardless of the type?

**Comment by YuriiMotov** [comment]
@khuloud0, see https://github.com/fastapi/fastapi/pull/14322

**Comment by vincent067** [comment]
Hey everyone! 👋 

I was looking into this same confusion today and found this issue. After digging around, I discovered that there's already an open PR #14322 by @ChaitanyaSai-Meka that addresses exactly this documentation problem.

The PR clarifies that HTTP query parameters are transmitted as strings, so clients can't actually send Python's  or JSON  directly. It adds helpful examples showing how to handle cases where you want to accept a "None" string or empty string as  in your route.

**Current status of the fix:**
- ✅ Reviewed by @YuriiMotov (maintainer)
- ✅ All feedback has been addressed
- ⏳ Waiting for final review by Sebastian
- ⚠️ Has merge conflicts that need resolving

For anyone blocked by this in the meantime, the workaround is:
1. Use  without  if the parameter is optional
2. If you need to accept literal "None" or empty strings as , handle that conversion inside your route or use Pydantic's  (v2 only)

@ChaitanyaSai-Meka - thanks for putting together this PR! The documentation clarification will definitely help a lot of people. Would you like help resolving the merge conflicts, or are you all set there? 😊

Hope this helps anyone else who stumbles upon this issue!

**Comment by khuloud0** [comment]
@YuriiMotov @vincent067 Thanks for the update @vincent067! It’s great to know that a PR (#14322) is already addressing this. I’ll follow the progress there to learn how the documentation will be clarified. Appreciate the workarounds you mentioned!

**Comment by ChaitanyaSai-Meka** [comment]
Hey @vincent067 , thnks a lot for th kind words and for pointing ths out! 
I just checked again,  it looks like there are no merge conflicts on the branch anymore (GitHub shows it can be cleanly merged).
Pls let me know if you’re seeing anything different on your end

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — I think this one is more of a docs/Q&A item than a FastAPI bug, and it can probably be closed.

The question asks how to "send `None`" to a query parameter declared as `q: Annotated[str | None, Query(min_length=3)] = ...`. The `...` (`Ellipsis`) default just means *required*; it doesn't add any "must-not-be-null" semantics on top of `str | None`. The real friction is at the HTTP layer:

- HTTP query strings have no native `null` literal — `?q=` sends an empty string, not `None`. `Optional[str]` in a query is "may be omitted from the URL", not "may be the JSON `null`". This is an HTTP/RFC 3986 constraint, not a FastAPI choice.
- If the goal is "the param must be present, but its value can be the empty string", that's just `Annotated[str, Query()] = ...` and the client sends `?q=`. FastAPI treats the empty string as a valid value.
- If the goal is "the client can explicitly send `null`", that needs to move to the request body (JSON `null` is representable) or to a header convention; query strings can't express it without a sentinel.

The `min_length=3` then conflicts with both `""` and any sentinel under 3 chars, which is probably what surprised the original poster — but again, that's a constraint they wrote, not a FastAPI bug.

Would you mind closing this in favour of the upstream discussion ([#12418](https://github.com/fastapi/fastapi/discussions/12418))? Happy to be wrong if I missed an angle of the question.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the HTTP/RFC reference and the FastAPI behaviour above were verified manually against the current `master` branch.)

**Comment by AayushShah-904** [comment]
Hi @Kludex and FastAPI team,

I would love to take this on as my first contribution to the project! 

I've reproduced the behavior locally. As noted, since query strings handle data as literal text, passing `?q=None` or `?q=null` processes them as string types rather than native Python `None` objects, and omitting the parameter entirely causes a `422 validation error` due to the required ellipsis (`...`).

**My Plan:**
1. I will update `docs/en/docs/tutorial/query-params-str-validations.md` under the `### Required, can be None` header with a clear warning callout block explaining this technical URL limitation.
2. I'll provide clear guidance for users that if a query parameter genuinely needs to be optional or omitted, dropping the `...` default in favor of an explicit `= None` default value is the proper path.
3. Verify that all existing unit tests in `tests/test_docs.py` pass perfectly.

**Comment by YuriiMotov** [comment]
@AayushShah-904, please, start from reviewing existing PRs linked to the issue (including closed) to ensure you are not going to create duplicate

**Comment by tiangolo** [comment]
We'll continue tracking this internally, I'll close this one to prevent more automated AI generated PRs that only add to our overhead. ☕ 

https://fastapi.tiangolo.com/contributing/#automated-code-and-ai

**Comment by WilliamDEdwards** [comment]
@tiangolo 

> We'll continue tracking this internally

Not very _open_ source?
